File: String/(E) Roman to Integer/code.java
# https://leetcode.com/problems/roman-to-integer/
class Solution {
    
     HashMap<Character,Integer> map = new HashMap<Character,Integer>();
    
    public Solution() {
        map.put('h',0);
        map.put('I', 1);
        map.put('V', 5);
        map.put('X', 10);
        map.put('L', 50);
        map.put('C', 100);
        map.put('D', 500);
        map.put('M', 1000);
    }
    
    public int romanToInt(String s) {
        int ret=0;
        Character prev='h';
        for (int i=0; i<s.length(); i++){
            Character t = s.charAt(i);
            if (prev.equals('h')) {
                prev=t;
            }
            else if ((prev.equals('I') && (t.equals('V') || t.equals('X'))) || (prev.equals('X') && (t.equals('L') || t.equals('C'))) || (prev.equals('C') && (t.equals('D') || t.equals('M')))) {
                ret+=(-1*map.get(prev)+map.get(t));
                prev='h';
            }
            else {
                ret+=map.get(prev);
                prev=t;
            }
        }
        
        ret+=map.get(prev);
        return ret;
    }
}

File: DP/Spiral Matrix/code.py
# https://leetcode.com/problems/spiral-matrix/
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        l=0
        u=0
        m=len(matrix)
        n=len(matrix[0])
        if m==0:
            return []
        elif n==1:
            ret=[]
            for i in range(len(matrix)):
                ret.append(matrix[i][0])
            return ret
        d=m-1
        r=n-1
        end=m*n
        i,j=0,0
        ret=[matrix[i][j]]
        count=1
        while(j+1<=r):
            # print("hi")
            while(j+1<=r):
                j+=1
                ret.append(matrix[i][j])
                count+=1
            if count==m*n: break
            u+=1
            while(i+1<=d):
                i+=1
                ret.append(matrix[i][j])
                count+=1
            if count==m*n: break
            r-=1
            while(j-1>=l):
                j-=1
                ret.append(matrix[i][j])
                count+=1
            if count==m*n: break
            d-=1
            while(i-1>=u):
                i-=1
                ret.append(matrix[i][j])
                count+=1
            if count==m*n: break
            l+=1
        return ret
